test/test.cc

```diff
@@ -6512,14 +6512,33 @@ TEST(TaskQueueTest, IncreaseAtomicInteger) {
       new ThreadPool{CPPHTTPLIB_THREAD_POOL_COUNT}};
 
   for (unsigned int i = 0; i < number_of_task; ++i) {
-    task_queue->enqueue(
+    bool queued = task_queue->enqueue(
         [&count] { count.fetch_add(1, std::memory_order_relaxed); });
+    EXPECT_TRUE(queued);
   }
 
   EXPECT_NO_THROW(task_queue->shutdown());
   EXPECT_EQ(number_of_task, count.load());
 }
 
+TEST(TaskQueueTest, IncreaseAtomicIntegerWithQueueLimit) {
+  static constexpr unsigned int number_of_task{1000000};
+  unsigned int queued_count{0};
+  std::atomic_uint count{0};
+  std::unique_ptr<TaskQueue> task_queue{
+      new ThreadPool{/*num_threads=*/1, /*qlimit=*/2}};
+
+  for (unsigned int i = 0; i < number_of_task; ++i) {
+    if (task_queue->enqueue(
+        [&count] { count.fetch_add(1, std::memory_order_relaxed); })) {
+      queued_count++;
+    }
+  }
+
+  EXPECT_NO_THROW(task_queue->shutdown());
+  EXPECT_EQ(queued_count, count.load());
+}
```

Comment:
Could you insert `EXPECT_TRUE(queued_count < number_of_task);` at line 6540?

Reply:
Thanks for the adjustment. But I suggested `<` instead of `<=`, so that we can confirm that `task_queue->enqueue` fails at least once.

Reply:
Although extremely unlikely, it is still possible that `enqueue` never fails.
The worker thread loop may be faster than the enqueue loop, in such a way that the latter will never find the queue full.

You can reproduce this condition by adding something like `std::this_thread::sleep_for(std::chrono::microseconds(100));` in the enqueue loop.

Reply:
What we can add, instead, is `EXPECT_TRUE(queued_count >= 2)`, because even in the worst case (worker thread starts after the enqueue loop) two enqueue operations will always succeed.

Reply:
To be honest, I would like to see something simpler which clearly shows that this edge case is properly handled. The original IncreaseAtomicInteger test isn't intended for this purpose, and I don't think it's a good idea to reuse it as base...

Reply:
Do you have any additional ideas?

To me this looks a very good test, because it will, in practice, heavily exercise both the cases (enqueue successful or not): on my machine I see about 15% of successful enqueue and 85% of failures. And it checks that the counter increments only when the enqueue is successful, thus validating the return value of enqueue.

Since the workers start on ThreadPool constructor (rather than with a separate `start()` method), in order to force an enqueue to fail I can only think of even more complex tests...

Reply:
I came up with the following simple edge case test.

```cpp
TEST(TaskQueueTest, MaxQueuedRequests) {
  std::unique_ptr<TaskQueue> task_queue{new ThreadPool{1, 2}};

  auto count = 8;
  while (count-- > 0) {
    task_queue->enqueue(
        [] { std::this_thread::sleep_for(std::chrono::milliseconds(50)); });
  }

  auto ret = task_queue->enqueue([] {});
  EXPECT_FALSE(ret);

  std::this_thread::sleep_for(std::chrono::milliseconds(100));

  ret = task_queue->enqueue([] {});
  EXPECT_TRUE(ret);

  EXPECT_NO_THROW(task_queue->shutdown());
}
```

I tested it 100 times with the following command, and confirmed that it worked successfully. Of course, there is a slight change to see it fails, but I think it's really rare and ok to put it in the unit test.

```
~/Projects/cpp-httplib/test$ make test && for i in {0..99}; do ./test --gtest_filter="*MaxQueued*"; done
```

Could you add the above the unit test case right below your `IncreaseAtomicIntegerWithQueueLimit`?

Also please run the clang-format on 'httplib.h' and 'test/test.cc'. Thanks for your help!

Reply:
clang-format done.

What about the same test idea you proposed, but with proper synchronization?

Reply:
Looks much nicer!

Reply:
Could take a look at the failure in 'build (macOS-latest)' on Github Actions?
https://github.com/yhirose/cpp-httplib/actions/runs/7309110277/job/19916387047?pr=1741

Reply:
There was a logic bug in the test. Now is fixed.

Reply:
Now I've run the test 10000 times and it always succeded.